### DISCUSSION TITLE
How to properly do OSSL_LIB_CTX initialization with custom elements

### DISCUSSION THREAD

**Discussion by RodriM11** [openssl/openssl#25495] [Q&A] [answered]
Hello! I am wondering how would it be recommended to instantiate custom ```EVP_RAND_CTX``` public and private instances within ```OSSL_LIB_CTX``` * elements (i.e., select not only DRBG_type, which can be done via ```RAND_set_DRBG_type```, but also _personalization strings_ and other params such as _strength_, _reseed_request_, _reseed_intervals_, etc...? An initial approach could be to just initialize those in a standard way, and then setting them via ```RAND_set0_public/private``` to the desired ```OSSL_LIB_CTX```*, but the documentation states that "it is recommended not to change the settings of these three instances directly. Instead, an application should change the default settings for new DRBG instances at initialization time, before creating additional threads." and certain attempts I've made in using the ```RAND_set0_public/private```  have indeed sometimes resulted in Segmentation Faults when freeing every element generated, if using multiple threads. 

My question would then be if there is a more established way of initiating private/public ```EVP_RAND_CTX``` * of a ```OSSL_LIB_CTX``` *, with all the attributes mentioned above, without employing the ```RAND_set0_``` interface.

**Comment by paulidale** [comment, marked as answer]
You shouldn't be setting _strength_ since that is a property inherent in the EVP_RAND.

Are you attempting to use the DRBGs supplied as part of OpenSSL with different parameters or are you installing a custom EVP_RAND?

For the former, let them create as usual and then call EVP_RAND_CTX_set_params to adjust the reseed requests and reseed time as required.  This doesn't cover the personalisation string, for which there is no method to change it currently.  If you must change this, you'll need to change the definition in `providers/implementations/rands/drbg_local.h`.  I don't see a complelling reason for changing this string.  It's not considered to be entropy input and will be attacker known -- i.e. any string is just as good as any other.

Build time changing of the personalisation string would be a matter of adding an `#ifndef DRBG_DEFAULT_PERS_STRING` wrapper around the aforementioned definition & documenting it in `INSTALL.md`.  This is a feature I expect would be accepted by the project if someone were to submit a pull request.

**Reply by RodriM11** [reply]
So, you mean letting them be created as usual, then access them through ```RAND_get0_``` and then use ```EVP_RAND_CTX_set_params``` over them to modify the setting, right?

**Reply by paulidale** [reply]
Exactly.  Or just call RAND_get0_ which will create them and then modify before use.
